test/run_test.py

```diff
@@ -444,6 +475,114 @@ def get_selected_tests(options):
     return selected_tests
 
 
+def test_impact_of_file(filename):
+    """Determine what class of impact this file has on test runs.
+
+    Possible values:
+        TORCH - torch python code
+        CAFFE2 - caffe2 python code
+        TEST - torch test code
+        UNKNOWN - may affect all tests
+        NONE - known to have no effect on test outcome
+    """
+    parts = filename.split(os.sep)
+    if parts[0] in ['.jenkins', '.circleci', 'docs', 'scripts'] or parts[-1] == 'run_test.py':
+        # HACK: ignore existence of CI config files in order to support testing of 
+        # target determination.
+        return 'NONE'
+    elif parts[0] == 'torch' and parts[-1].endswith('.py'):
+        return 'TORCH'
+    elif parts[0] == 'caffe2' and parts[-1].endswith('.py'):
+        return 'CAFFE2'
+    elif parts[0] == 'test' and parts[-1].endswith('.py'):
+        return 'TEST'
+
+    return 'UNKNOWN'
+
+
+def log_test_reason(file_type, filename, test, options):
+    if options.verbose:
+        print_to_stderr(
+            'Determination found {} file {} -- running {}'.format(
+                file_type,
+                filename, 
+                test,
+            )
+        )
+
+
+def get_dep_modules(test):
+    repo_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    test_location = os.path.join(repo_root, 'test', test + '.py')
+    finder = modulefinder.ModuleFinder(
+        # Ideally exclude all third party modules, to speed up calculation.
```

Comment:
Excludes seems like the wrong way to do this, on multiple accounts: (1) it appears that ModuleFinder is doing an exact match for entries in excludes, which means that if I directly import a second level module I'll zoom off into the netherworld of those modules as well, (2) a **whitelist** seems like a far more appropriate method of deciding whether or not to consider modules, since it's easy to say we only care about `torch` modules. But it doesn't see like ModuleFinder permits predicates, so you'll have to either monkeypatch the class or reimplement it yourself. I wonder if this is why ModuleFinder takes a long time; have you run it under verbose mode and looked at what it's traversing?

Reply:
Another problem with a whitelist approach is that imports within the `test/` path do not share a common "test." namespace -- so you'd have to do some magic to even produce the whitelist.